src/infra/core/src/repos/metadata_block_repository_impl.rs

```diff
@@ -0,0 +1,125 @@
+    InsertBlockResult,
+    InsertOpts,
+    MetadataBlockRepository,
+    ObjectRepository,
+};
+use opendatafabric::serde::flatbuffers::{
+    FlatbuffersMetadataBlockDeserializer,
+    FlatbuffersMetadataBlockSerializer,
+};
+use opendatafabric::serde::{Error, MetadataBlockDeserializer, MetadataBlockSerializer};
+use opendatafabric::{MetadataBlock, Multihash};
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+pub struct MetadataBlockRepositoryImpl<ObjRepo> {
+    obj_repo: ObjRepo,
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+impl<ObjRepo> MetadataBlockRepositoryImpl<ObjRepo>
+where
+    ObjRepo: ObjectRepository + Sync + Send,
+{
+    pub fn new(obj_repo: ObjRepo) -> Self {
+        Self { obj_repo }
+    }
+
+    pub fn deserialize_metadata_block(
+        hash: &Multihash,
+        block_bytes: &[u8],
+    ) -> Result<MetadataBlock, GetBlockError> {
+        FlatbuffersMetadataBlockDeserializer
+            .read_manifest(block_bytes)
+            .map_err(|e| match e {
+                Error::UnsupportedVersion { .. } => {
+                    GetBlockError::BlockVersion(BlockVersionError {
+                        hash: hash.clone(),
+                        source: e.into(),
+                    })
+                }
+                _ => GetBlockError::BlockMalformed(BlockMalformedError {
+                    hash: hash.clone(),
+                    source: e.into(),
+                }),
+            })
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait]
+impl<ObjRepo> MetadataBlockRepository for MetadataBlockRepositoryImpl<ObjRepo>
+where
+    ObjRepo: ObjectRepository + Sync + Send,
+{
+    async fn contains_block(&self, hash: &Multihash) -> Result<bool, ContainsBlockError> {
+        self.obj_repo.contains(hash).await.map_err(Into::into)
+    }
+
+    async fn get_block(&self, hash: &Multihash) -> Result<MetadataBlock, GetBlockError> {
+        let block_data = self.obj_repo.get_bytes(hash).await?;
+
+        Self::deserialize_metadata_block(hash, &block_data)
+    }
+
+    async fn insert_block<'a>(
+        &'a self,
+        block: &MetadataBlock,
+        options: InsertOpts<'a>,
+    ) -> Result<InsertBlockResult, InsertBlockError> {
+        let block_data = FlatbuffersMetadataBlockSerializer
+            .write_manifest(block)
+            .int_err()?;
+
+        self.obj_repo
+            .insert_bytes(&block_data, options)
+            .await
+            .map(Into::into)
+            .map_err(Into::into)
+    }
+
+    fn as_object_repo(&self) -> &dyn ObjectRepository {
+        &self.obj_repo
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+// Helpers
+/////////////////////////////////////////////////////////////////////////////////////////
+
+pub trait MetadataBlockRepositoryExt<ObjRepo>
```

Comment:
Используется для следующего фрагмента:
```rust
impl<WrappedRepo, ObjRepo> MetadataBlockRepositoryCachingInMem<WrappedRepo, ObjRepo>
where
    WrappedRepo: MetadataBlockRepositoryExt<ObjRepo> + Sync + Send,
    ObjRepo: ObjectRepository + Sync + Send,
{
    pub fn new(obj_repo: ObjRepo) -> Self {
        let wrapped = WrappedRepo::new(obj_repo); // <---

        MetadataBlockRepositoryCachingInMem::new_wrapped(wrapped)
    }
...
    ```

Reply:
I think this brings more complexity than it solves.

If in `dataset_factory.rs` instead of:
```rust
MetadataBlockRepositoryImplWithCache::new(ObjectRepositoryLocalFSSha3::new(
  layout.blocks_dir,
)),
```

we created instance as:
```rust
MetadataBlockRepositoryCachingInMem::new(
  MetadataBlockRepositoryImpl::new(
    ObjectRepositoryLocalFSSha3::new(layout.blocks_dir)
  )
)
```
It would:
- make `MetadataBlockRepositoryExt` trait not necessary
- make `pub type MetadataBlockRepositoryImplWithCache` alias not necessary
- also I think it would make compositional nature of cache more explicit - I find it really like to see how objects compose upon creation
- and also it might make it easier to introduce a per-process / redis cache in future - we will just slot in another layer of caching under InMem - in current code we won't be able to add another wrapper in the middle because `MetadataBlockRepositoryImpl` is hidden from the construction process

Reply:
Done